### ISSUE TITLE
Since AtomicState can implement the full interface of State, why not have it be the default State?

### ISSUE THREAD

**Issue by lehmacdj** [polysemy-research/polysemy#362]
It seems like atomic state is more general, therefore it would allow code to be interpreted in more useful ways.

My proposal is simple: reimplement State the same way that AtomicState is implemented. This means that we would have:
```haskell
data State m a where
    AtomicState :: (s -> (s, a)) -> AtomicState s m a 
    AtomicGet :: AtomicState s m s
```
The implementation of the actions would be the same as the ones from Polysemy.AtomicState, essentially:
```haskell
get = atomicGet
put = atomicPut
modify = atomicModify

etc.
```
We would then be able to provide the atomic interpretations for this module in addition to the existing ones for the State effect.

For backwards compatibility we can keep the Polysemy.AtomicState module around. We can also keep around the current State effect as NonAtomicState if performance implications would cause someone to prefer that module over one that is compatible with atomicity.

Advantages:
- Code written with the State effect could be interpreted such that all of the actions are interpreted atomically.
- Less verbose code using atomic state

Drawbacks:
- Performance may be worse. I'm not knowledgeable enough about Haskell performs to make a good judgement here.
- Code written for State might not expect code to be run expecting atomic behavior. This might result in subtle concurrency bugs if people expect code that is interpretable in an atomic way to be written with concurrency in mind. Note: that this would not cause the code to violate the promises of atomicity, but merely it would still require care to reinterpret code written using the State effect atomically.

Assumptions:
- The performance of AtomicState and State are close enough that swapping out the current implementation of `put` with essentially `atomicPut` is small enough that this api change is possible.

**TLDR**: As I wrote this I pretty much decided that the second drawback is probably large enough that it makes this not worth doing. Having to use AtomicState instead of State to get atomic interpretations is probably good as *syntactic salt* as a means for having people be more careful when writing concurrent code. However I'm still interested in if you (people working on polysemy have considered this design and if so what you're reasons for not going with it are).

**Comment by KingoftheHomeless** [comment]
Your TLDR is on the mark. I've considered this before, but the second drawback sinks it, since `State` isn't required to be atomic. `AtomicState` explicitly signals the need to avoid race conditions, and any combinatorial code that works with `AtomicState` can be expected to keep that in mind. If `AtomicState` and `State` were merged, we'd lose out on that; so no matter what, we need to have a separate `AtomicState` effect.

We _could_ have `State` use the same basic interface as `AtomicState`, (replacing `Put` with `State`) but there's nothing really to be gained from doing that.

**Comment by tek** [comment]
the doc for `State` states (:stuck_out_tongue: ) that it's supposed to be more abstract ("implemented as http calls") so that seems like a justification for leaving out atomicity

**Comment by lehmacdj** [comment]
But admitting atomic interpretations is also more abstract as it admits more possible interpretations. 

I agree that the syntactic salt argument is good enough though that it isn't worth it though.

Thanks for confirming what I realized myself!
